ops/ccs-ops-misc/synthetic-data/scripts/synthea-automation/s3_utilities.py

```diff
@@ -0,0 +1,358 @@
+#
+# Param: target_dir : unix filesystem directory where downloaded S3 file(s) are written to.
+# Raises a python exception if failure to download file.
+def download_synthea_files(target_dir):
+    for fn in code_map_files:
+        output_fn = target_dir + fn
+        try:
+            s3_client.download_file(mitre_synthea_bucket, fn, output_fn)
+        except botocore.exceptions.ClientError as e:
+            if e.response['Error']['Code'] == "404":
+                print(f"The object does not exist: {fn}")
+            else:
+                raise
+
+# Function to download proprietary Mitre shell script files from an S3 bucket; the scripts
+# drive the synthea generation process. The scripts are listed in an array of filenames
+# so this function loops over names in the code_script_files array. Since the files will
+# be run as shell script(s), unix file permissions are set to enable them to be executed.
+#
+# Param: target_dir : unix filesystem directory where downloaded S3 file(s) are written to.
+# Raises a python exception if failure to download file.
+def download_synthea_scripts(target_dir):
+    for fn in code_script_files:
+        output_fn = target_dir + fn
+        try:
+            s3_client.download_file(mitre_synthea_bucket, fn, output_fn)
+            os.chmod(os.fspath(output_fn), 0o744)
+        except botocore.exceptions.ClientError as e:
+            if e.response['Error']['Code'] == "404":
+                print(f"The object does not exist: {fn}")
+            else:
+                raise
+
+# Function to download the Mitre BFD end_state.properties file from an S3 bucket;
+# the file contains meta information of the previous run of the synthea data generation
+# process. Maintaining state between runs allows synthea generation to not create
+# redudnant data. 
+#
+# Param: target_dir : unix filesystem directory where downloaded S3 end_state.properties
+#        file is written to.
+# Raises a python exception if failure to download file.
+def download_end_state_props_file(target_dir):
+    base_name = os.path.basename(end_state_props_file)
+    output_fn = target_dir + base_name
+    try:
+        s3_client.download_file(mitre_synthea_bucket, end_state_props_file, output_fn)
+    except botocore.exceptions.ClientError as e:
+        if e.response['Error']['Code'] == "404":
+            print("The object does not exist.")
+        else:
+            raise
+
+# Function to upload a newly generated end_state.properties file from the synthea
+# run, to the Mitre BFD S3 bucket. We'll also upload the _orig version of that same
+# file which was created based on the previous synthea run. 
+#
+# Param: synthea_output_dir : unix filesystem directory where synthea writes the end_state.properties to.
+# Raises a python exception if failure to upload file.
+def upload_end_state_props_file(synthea_output_dir, bfd_dir):
+    # Mitre FQN for storing end_state.properties file
```

Comment:
comment can probably be removed, I think whatever was here got moved up

Reply:
yea, there were a few cut-paste errors that I saw when going though things again. fixed.